### PR TITLE
Use params keyword with ActionController::TestCase

### LINKED ISSUE #905: ActionController::TestCase needs keyword args
```
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
```

Occurs hundreds of times in test suite.

### LINKED ISSUE #907: :nothing option deprecated
```
DEPRECATION WARNING: `:nothing` option is deprecated and will be removed in Rails 5.1. 
Use `head` method to respond with empty response body. 
(called from block (4 levels) in <top (required)> at
 /Users/atz/repos/curation_concerns/spec/controllers/downloads_controller_spec.rb:10)
```

### LINKED ISSUE #910: xhr deprecated
```
DEPRECATION WARNING: 
xhr and xml_http_request methods are deprecated in favor of `get :index, xhr: true` and `post :create, xhr: true`
(called from block (5 levels) in <top (required)> at
 .../spec/controllers/curation_concerns/file_sets_controller_spec.rb:98)
```

### PR DESCRIPTION
**Pull request by atz**
Fixes #910
Fixes #907 
Largely fixes #905.  Remaining TestCase warnings are from a helper class.

Saves hundreds of lines of deprecation warnings per CI run.

Converts `xhr` calls to their corresponding HTTP method with `xhr: true`.  (#910)

### CODE REVIEW

**Comment by jcoyne** [comment]
I suspect this will break for Rails 4, but you could copy this from Blacklight, which will fix it: https://github.com/projectblacklight/blacklight/commit/59436b48828c96a013a207fab079d82d4bc2c793#diff-93830fa29d616f7c87903d08b5b1b29aR70

**Comment by atz** [comment]
I thought the `params` part is old.  You think the `xhr: true` will break it?

**Comment by atz** [comment]
Yep, you're right about the break in Rails 4.  Thx for the tip, will fold that in.

**Comment by atz** [comment]
You know, for how fugly the specs are and the _dozens_ of rubocop exceptions CC has adopted, breaking the build on an extra line break seems pretty damn rich.

**Comment by jcoyne** [comment]
@atz create an executable (chmod 755) file in `.git/hooks/pre-commit`
```
#!/bin/sh

exec bundle exec rubocop
```

Then you'll never have that problem again.

**Comment by atz** [comment]
It was from your lines I copied over from Sufia, for the record.

**Comment by atz** [comment]
@jcoyne Any feedback on why the backport workaround isn't satisfying the routes under rails 4?
https://travis-ci.org/projecthydra/curation_concerns/jobs/152870163

Did I miss pulling something in?  Still works fine on rails 5.  Ah... looks like it is from `patch` requests that didn't get intercepted.

**Comment by atz** [comment]
Ok, it's down to 4x routing error:
```
ActionController::UrlGenerationError:
       No route matches {:action=>"show", :controller=>"curation_concerns/file_sets"}
```
And one `ActionController::UnknownFormat`.

https://travis-ci.org/projecthydra/curation_concerns/jobs/153150727

**Comment by atz** [comment]
```ruby
respond_to do |format|
  format.html do
    if request.xhr?
      render 'jq_upload', formats: 'json', content_type: 'text/html'
    else
      redirect_to [main_app, curation_concern.parent]
    end
  end
  format.json do
    render 'jq_upload', status: :created, location: polymorphic_path([main_app, curation_concern])
  end
end
```

XHR... requests html, give it json with html mime-type?  Is that the logic I'm seeing there?

**Comment by jcoyne** [comment]
@atz yeah, it doesn't make a lot of sense, but I believe it's due to one of the javascript libraries/browsers not setting the Accept type correctly on XHR requests.

**Comment by atz** [comment]
Probably deserves a comment to that effect, at least.  Would be nice to know to be able to rip it out when that lib is updated/replaced/removed.

**Comment by atz** [comment]
Got it down to just that one failure.

**Comment by jcoyne** [comment]
@atz one thing the old XHR method did, that our method is not doing is set the HTTP_ACCEPT header as seen here:
https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_controller/test_case.rb#L549

Perhaps you can modify that support file to add that line?

**Comment by jcoyne** [comment]
Or even change it to:
```ruby
if rest[:xhr]
  xml_http_request(:post, action, rest[:params])
else
 super(action, rest[:params])
end
```

**Comment by atz** [comment]
Yeah, the extra ENV var does it.  Seeing as HTTP_ACCEPT is so critical to server response, I'm not surprised that leaving it out would break something.  I collapsed the backports file down to 12 lines with the changes.  We probably want to use that version in blacklight, too, since it adds that ENV and `patch` support.

Thanks for the insights.

File: spec/support/backport_test.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Backport the Rails 5 controller test methods to Rails 4
+module BackportTest
+  [:delete, :get, :post, :put, :patch].each do |http_action|
+    define_method(http_action) do |*args|
+      (action, rest) = *args
+      rest ||= {}
+      @request.env['HTTP_X_REQUESTED_WITH'] = 'XMLHttpRequest' if rest[:xhr]
+      @request.env['HTTP_ACCEPT'] ||= [Mime::JS, Mime::HTML, Mime::XML, 'text/xml', Mime::ALL].join(', ')
```
Comment (jcoyne) [inline spec/support/backport_test.rb]: I think this should only be set `if rest[:xhr]`

Comment (atz) [inline spec/support/backport_test.rb]: Probably doesn't matter since all likely targets are covered, literally including Mime::ALL, but sure.

**Pull request merged**